### PR TITLE
Add check whether k8s using 3rd jwt or not.

### PR DESCRIPTION
**Pull request by vpn-soft**
https://istio.io/latest/docs/ops/best-practices/security/#configure-third-party-service-account-tokens

### CODE REVIEW

File: pkg/k8s/client.go [hunk 1]
```diff
@@ -229,6 +248,52 @@ func (c *Client) scanAll() {
...
+	if strings.Compare(ns, constants.IstioSystemNamespace) != 0 {
+		return nil
+	}
+	configMap, err := getK8SConfigMapFunc(client, "istio-sidecar-injector")
+	if err != nil {
+		log.Errorf("Error while fetching k8s configMap details :: Error: %v\n", err)
+		return nil
+	}
+	var values map[string]interface{}
+	err = json.Unmarshal([]byte(configMap.Data["values"]), &values)
```
Comment (jcchavezs) [inline pkg/k8s/client.go:261]: Shall we check `"values"` key exists?

Comment (vpn-soft) [reply pkg/k8s/client.go]: Done

**Review by jcchavezs — APPROVED** [review]

File: pkg/k8s/client.go [hunk 2]
```diff
@@ -229,6 +248,57 @@ func (c *Client) scanAll() {
 	c.mu.Unlock()
 }
 
+func checkJWTPolicy(ns string, client *Client) (err error) {
```
Comment (incfly) [inline pkg/k8s/client.go:251]: remove naming `error` as `err`. just `func checkJWTPolicy(ns string, client *Client) error {`

https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#_avoid_named_return_values

File: pkg/k8s/client.go [hunk 3]
```diff
@@ -229,6 +248,57 @@ func (c *Client) scanAll() {
...
+	if ns != constants.IstioSystemNamespace {
+		return nil
+	}
+	configMap, err := getK8SConfigMapFunc(client, "istio-sidecar-injector")
```
Comment (incfly) [inline pkg/k8s/client.go:255]: I understand why we use a `getK8SConfigMapFunc` for this. It's to make unit testable. However I think:

- We should avoid global variables. Changing global variable in different places makes code fragile.
- We can defer the whole k8s apiserver dependency mock later.

How to achieve that? We can split this function into:


```golang
// focus on getting the config map object.
func retrieveConfigMap(client *Client) (*corev1.ConfigMap, error) {
}

// focus on parsing and extracting .values.global.jwtPolicy values from the data.
func checkUses3rdPartyJWT(configMap *corev1.ConfigMap) error {
}
```

And in the unit test, you can only test `checkUses3rdPartyJWT` by providing some sample config map object (it's easy to prepare an inlined YAML string and do conversion to the k8s object. let me know if you need some help on that.)

File: pkg/k8s/client.go [hunk 4]
```diff
@@ -229,6 +248,57 @@ func (c *Client) scanAll() {
...
+	if err != nil {
+		log.Errorf("Error while fetching k8s configMap details :: Error: %v\n", err)
+		return nil
+	}
+	var values map[string]interface{}
+	data, ok := configMap.Data["values"]
+	if !ok {
+		log.Errorf("returning, since no values configured in config map")
```
Comment (incfly) [inline pkg/k8s/client.go:263]: we should not log things to start with "returning". That's helpful for development, but not useful for debugging or users.
please reword it to make it a self-contained sentence, "unable to determine using 3rd party jwt or not: ....."

File: pkg/k8s/client.go [hunk 5]
```diff
@@ -229,6 +248,57 @@ func (c *Client) scanAll() {
...
+		return nil
+	}
+	err = json.Unmarshal([]byte(data), &values)
+	if err != nil {
+		log.Errorf("Unable to convert values data of configMap into map :: Error: %v\n", err)
```
Comment (incfly) [inline pkg/k8s/client.go:268]: Change " :: Error " to ": %v".

Same for all log statement below.

File: pkg/k8s/client.go [hunk 6]
```diff
@@ -229,6 +248,57 @@ func (c *Client) scanAll() {
...
+		return nil
```
Comment (incfly) [inline pkg/k8s/client.go:269]: we should return error in case we runs into error due to data inconsistency in the configmap.
we can use two errors to distinguish.

- unknown error, unable to determine whether it's 3rd party jwt or not.
- known error, we know, but the value is configured as 3rd party jwt enabled.

you can declare some constant/vars in the beginning of this file, like code in the parser lib, for the error about destination rule, authz policy, etc. `unableDetermine3rdPartyJwt = errors.New(...)`.

File: pkg/k8s/client_test.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+import (
+	"testing"
+
+	"istio.io/istio/pkg/config/constants"
+	corev1 "k8s.io/api/core/v1"
+)
+
+func TestCheckJWTPolicies(t *testing.T) {
+	tempGetCMFunc := getK8SConfigMap
```
Comment (incfly) [inline pkg/k8s/client_test.go:25]: Let's change the test to data driven test. https://dave.cheney.net/2019/05/07/prefer-table-driven-tests
search for "Introducing table driven tests" section.
you can also find many examples in other unit test.

File: pkg/k8s/client.go [hunk 8]
```diff
@@ -244,57 +247,61 @@ func (c *Client) scanAll() {
...
 	if err != nil {
-		log.Errorf("Error while fetching k8s configMap details :: Error: %v\n", err)
-		return nil
+		log.Errorf("Unable to fetch k8s config map : %v\n", err)
+		return errors.New(JWTPolicyUnknownError)
 	}
+	return checkUses3rdPartyJWT(configMap)
+}
+
+func retrieveConfigMap(client *Client) (*corev1.ConfigMap, error) {
```
Comment (incfly) [inline pkg/k8s/client.go:262]: be more specific? `retrieveSideCarInjectorConfigMap`?

File: pkg/k8s/client.go [hunk 9]
```diff
@@ -42,9 +42,8 @@ import (
 )
 
 var (
-	// getK8SConfigMapFunc function as a variable to mock k8s functions for unit testing, NOTE: temporarily placed till k8s client
-	// api mocks will be in place
-	getK8SConfigMapFunc = getK8SConfigMap
+	JWTPolicyUnknownError = "JWT policy: unable to determine whether it's 3rd party jwt or not"
```
Comment (incfly) [inline pkg/k8s/client.go:45]: we could just type them as error. `jwtPolicyUnknownError = errors.New("....")`.

also probably just lower case it to make it private. we don't have a need to make it accessible outside this package yet.

**Review by incfly — APPROVED** [review]
looks good, just a few nitpicks

File: pkg/k8s/client.go (hunk 2 above)
Comment (vpn-soft) [reply pkg/k8s/client.go]: Done

File: pkg/k8s/client.go (hunk 3 above)
Comment (vpn-soft) [reply pkg/k8s/client.go]: Done

File: pkg/k8s/client.go (hunk 4 above)
Comment (vpn-soft) [reply pkg/k8s/client.go]: Done

File: pkg/k8s/client.go (hunk 5 above)
Comment (vpn-soft) [reply pkg/k8s/client.go]: Done

File: pkg/k8s/client.go (hunk 6 above)
Comment (vpn-soft) [reply pkg/k8s/client.go]: Done

File: pkg/k8s/client_test.go (hunk 7 above)
Comment (vpn-soft) [reply pkg/k8s/client_test.go]: Done

File: pkg/k8s/client.go (hunk 8 above)
Comment (vpn-soft) [reply pkg/k8s/client.go]: Done

File: pkg/k8s/client.go (hunk 9 above)
Comment (vpn-soft) [reply pkg/k8s/client.go]: Done

**Review by incfly — APPROVED** [review]

File: pkg/k8s/client.go [hunk 10]
```diff
@@ -193,6 +200,7 @@ func (c *Client) scanAll() {
 	// Iterate namespaces.
 	configs := []*istioconfig.Config{}
 	namespaces := c.nsInformer.GetIndexer().List()
+	var jwtPolicyError error = nil
```
Comment (jcchavezs) [inline pkg/k8s/client.go:203]: I think you could just not use the `= nil`
